src/main/scala/uk/gov/nationalarchives/tdr/api/service/ValidateFileMetadataService.scala

```diff
@@ -0,0 +1,113 @@
+package uk.gov.nationalarchives.tdr.api.service
+
+import uk.gov.nationalarchives.Tables.{FilemetadataRow, FilestatusRow}
+import uk.gov.nationalarchives.tdr.api.db.repository.{FileMetadataRepository, FileStatusRepository}
+import uk.gov.nationalarchives.tdr.api.graphql.fields.CustomMetadataFields.{CustomMetadataField, CustomMetadataValues}
+import uk.gov.nationalarchives.tdr.api.service.FileStatusService._
+
+import java.sql.Timestamp
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
+
+class ValidateFileMetadataService(
+    customMetadataService: CustomMetadataPropertiesService,
+    fileMetadataRepository: FileMetadataRepository,
+    fileStatusRepository: FileStatusRepository,
+    timeSource: TimeSource
+)(implicit val ec: ExecutionContext) {
+
+  def toPropertyNames(fields: Seq[CustomMetadataField]): Set[String] = fields.map(_.name).toSet
+
+  def toAdditionalMetadataFieldGroups(fields: Seq[CustomMetadataField]): Seq[FieldGroup] = {
+    val closureFields = fields.filter(f => f.propertyGroup.contains("MandatoryClosure") || f.propertyGroup.contains("OptionalClosure"))
+    val descriptiveFields = fields.filter(f => f.propertyGroup.contains("OptionalMetadata"))
+    Seq(FieldGroup(ClosureMetadata, closureFields), FieldGroup(DescriptiveMetadata, descriptiveFields))
+  }
+
+  def toValueDependenciesGroups(field: CustomMetadataField): Seq[FieldGroup] = {
+    val values: List[CustomMetadataValues] = field.values
+    values.map(v => {
+      FieldGroup(v.value, v.dependencies)
+    })
+  }
+
+  def validateAdditionalMetadata(fileIds: Set[UUID], consignmentId: UUID, propertiesToValidate: Set[String]): Future[List[FilestatusRow]] = {
+    for {
+      customMetadataFields <- customMetadataService.getCustomMetadata
+      existingMetadataProperties: Seq[FilemetadataRow] <- fileMetadataRepository.getFileMetadata(consignmentId, Some(fileIds), Some(toPropertyNames(customMetadataFields)))
+    } yield {
+      val additionalMetadataFieldGroups: Seq[FieldGroup] = toAdditionalMetadataFieldGroups(customMetadataFields)
+      val additionalMetadataPropertyNames: Set[String] = additionalMetadataFieldGroups.flatMap(g => toPropertyNames(g.fields)).toSet
+
+      if (!propertiesToValidate.subsetOf(additionalMetadataPropertyNames)) {
+        List()
+      } else {
+        val additionalMetadataStatuses = {
+          additionalMetadataFieldGroups
+            .flatMap(group => {
+              val states = group.fields.flatMap(f => checkPropertyState(fileIds, f, existingMetadataProperties))
+              val filesWithNoAdditionalMetadataStatuses = fileIds
+                .filter(id => !states.map(_.fileId).contains(id))
+                .map(id => {
+                  FilestatusRow(UUID.randomUUID(), id, group.groupName, NotEntered, Timestamp.from(timeSource.now))
```

Comment:
This doesn't really matter because it's only there for the tests really but the other services have a `UUIDSource` passed in the constructor and they use that to get a new UUID. There's another `randomUUID` on line 64. Feel free to leave it if it causes problems though.